Add Azure Language Services provider

psorensen opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

This PR adds Azure Language Services as a provider and adds it to the Excerpt Generation feature.

Closes #159 

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->
Testing Steps:
1. Create new Azure resource for Language Services
2. Enter Credentials into the Excerpt Generation feature screen
3. Confirm valid credentials allow feature to be enabled
4. Test 'Generate Excerpt' tag for allowed post types.

### Changelog Entry

> Added - Azure Language Services provider with Excerpt Generation.

### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @psorensen @cadic @jeffpaul @Sidsector9 @dkotter 

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [x] All new and existing tests pass.

psorensen committed:
initialize Azure Language provider

psorensen committed:
clean up settings and add auth check

psorensen committed:
add debug info

psorensen committed:
avoid confusion by disabling feature if not authenticated

psorensen committed:
add excerpt generation

psorensen committed:
fix filter doc

psorensen commented:
@dkotter How should I handle the VIP flag for `wp_remote_get` usage? Should I conditionally use it if defined, or ignore this rule?

dkotter commented:
> @dkotter How should I handle the VIP flag for `wp_remote_get` usage? Should I conditionally use it if defined, or ignore this rule?

So we've not been consistent with this so far in the codebase. We've basically done both, where we use `vip_safe_wp_remote_get` if it exists (see [here](https://github.com/10up/classifai/blob/develop/includes/Classifai/Providers/Azure/Speech.php#L216)) and otherwise use `wp_remote_get` and in other places we just use `wp_remote_get` and add an ignore statement.

Personally I'd prefer we go with the first approach, as this is technically a VIP approved plugin so it's nice to match their standards. We have had issues though a few times with timeouts, as some of these services take a bit to return. `vip_safe_wp_remote_get` only allows you to set a timeout of 3 seconds so if we need higher than that here, we'll want to stick with `wp_remote_get` only

psorensen committed:
add helper function to utilize vip remote get

psorensen commented:
sounds good @dkotter - I've included a helper function to facilitate a refactor of all the existing usage of `wp_remote_get`

psorensen committed:
Merge branch 'develop' into feature/azure-language-service

jeffpaul commented:
@psorensen is this still WIP or can this formally go through review/merge?

psorensen commented:
@jeffpaul Good to review - removed the WIP from the title

Sidsector9 commented on includes/Classifai/Providers/Azure/Language.php:
```diff
@@ -0,0 +1,391 @@
+		$endpoint = trailingslashit( $url ) . '/text/analytics/v3.1/languages';
+		$endpoint = add_query_arg( 'api-version', static::API_VERSION, $endpoint );
+
+		$request = safe_wp_remote_get(
```
This is throwing error. Should it use `wp_remote_post()` instead?

psorensen replied:
~~See this [comment](https://github.com/10up/classifai/pull/786#issuecomment-2269844794) and the follow-up comments for context. What error are you seeing?~~

Sidsector9 replied:
I am seeing the following when trying to connect to the service:

<img width="814" alt="Screenshot 2024-09-06 at 2 10 22 PM" src="https://github.com/user-attachments/assets/0cf92416-6b4c-4057-ac39-572c15a01676">

Sidsector9 replied:
Here's the neater version:

```
[06-Sep-2024 08:40:15 UTC] PHP Fatal error:  Uncaught TypeError: http_build_query(): Argument #1 ($data) must be of type array, string given in /Users/work/Sites/classifai/app/public/wp-includes/Requests/src/Transport/Curl.php:578
Stack trace:
#0 /Users/work/Sites/classifai/app/public/wp-includes/Requests/src/Transport/Curl.php(578): http_build_query('{"documents": [...', '', '&')
#1 /Users/work/Sites/classifai/app/public/wp-includes/Requests/src/Transport/Curl.php(391): WpOrg\Requests\Transport\Curl::format_get('https://classif...', '{"documents": [...')
#2 /Users/work/Sites/classifai/app/public/wp-includes/Requests/src/Transport/Curl.php(171): WpOrg\Requests\Transport\Curl->setup_handle('https://classif...', Array, '{"documents": [...', Array)
#3 /Users/work/Sites/classifai/app/public/wp-includes/Requests/src/Requests.php(469): WpOrg\Requests\Transport\Curl->request('https://classif...', Array, '{"documents": [...', Array)
#4 /Users/work/Sites/classifai/app/public/wp-includes/class-wp-http.php(397): WpOrg\Requests\Requests::request('https://classif...', Array, '{"documents": [...', 'GET', Array)
#5 /Users/work/Sites/classifai/app/public/wp-includes/class-wp-http.php(638): WP_Http->request('https://classif...', Array)
#6 /Users/work/Sites/classifai/app/public/wp-includes/http.php(184): WP_Http->get('https://classif...', Array)
#7 /Users/work/Sites/classifai/app/public/wp-content/plugins/classifai/includes/Classifai/Helpers.php(686): wp_remote_get('https://classif...', Array)
#8 /Users/work/Sites/classifai/app/public/wp-content/plugins/classifai/includes/Classifai/Providers/Azure/Language.php(164): Classifai\safe_wp_remote_get('https://classif...', Array)
#9 /Users/work/Sites/classifai/app/public/wp-content/plugins/classifai/includes/Classifai/Providers/Azure/Language.php(122): Classifai\Providers\Azure\Language->authenticate_credentials('https://classif...', '95193e52eec647f...')
#10 /Users/work/Sites/classifai/app/public/wp-content/plugins/classifai/includes/Classifai/Features/Feature.php(258): Classifai\Providers\Azure\Language->sanitize_settings(Array)
#11 /Users/work/Sites/classifai/app/public/wp-includes/class-wp-hook.php(326): Classifai\Features\Feature->sanitize_settings(Array)
#12 /Users/work/Sites/classifai/app/public/wp-includes/plugin.php(205): WP_Hook->apply_filters(Array, Array)
#13 /Users/work/Sites/classifai/app/public/wp-includes/formatting.php(5115): apply_filters('sanitize_option...', Array, 'classifai_featu...', Array)
#14 /Users/work/Sites/classifai/app/public/wp-includes/option.php(866): sanitize_option('classifai_featu...', Array)
#15 /Users/work/Sites/classifai/app/public/wp-admin/options.php(344): update_option('classifai_featu...', Array)
#16 {main}
  thrown in /Users/work/Sites/classifai/app/public/wp-includes/Requests/src/Transport/Curl.php on line 578

```

psorensen replied:
> Should it use wp_remote_post() instead?

Ignore my previous comment. Yes it should. 😄  Updated

Sidsector9 commented on includes/Classifai/Providers/Azure/Language.php:
```diff
@@ -0,0 +1,391 @@
+			}
+
+			while ( 'succeeded' !== $response->status ) {
+				sleep( .5 );
```
A downside of this approach is that if you navigate away from the page, the process will quit.
We can use Action Scheduler for background processing which we're using for the Classification feature. (#779)

@dkotter do you have any thoughts on this?

dkotter replied:
I think the problem here just in quickly looking at the code is that this method gets called from a REST request. The easy fix is what we're doing here, just pausing and retrying until we get a successful response.

I do wonder what happens if the response never succeeds? Would this code just be stuck in this loop?

Action Scheduler would be a nicer fix here but would be more complicated with how everything else is set up in ClassifAI. Basically when the `Generate Excerpt` button is clicked, the initial request would trigger Action Scheduler and then on the javascript side, we would need to continue polling until the request is done (and show the loading indicator the entire time).

That's doable but doesn't match how we handle things elsewhere so would need to think through how best to achieve that, especially in a way that would work for other Features and Providers in the future

psorensen replied:
@dkotter @Sidsector9  Knowing the action scheduler would be a heavier lift, how about a simple counter for the short term? See
* https://github.com/10up/classifai/pull/786/commits/a34e4f20cb5a313014860793c66fd94095ab4605
* https://github.com/10up/classifai/pull/786/commits/656b5454d562d1d33d68072fd74af66fbe2c6b66

Sidsector9 commented on includes/Classifai/Providers/Azure/Language.php:
```diff
@@ -0,0 +1,391 @@
+						 * @param string $summary_length The summary length.
+						 * @return string
+						 */
+						'summaryLength' => apply_filters( 'classifai_azure_language_summary_length', 'oneSentence' ),
```
We'll need to format the doc block

```php
/**
 * Filter the summary length.
 * Possible values are 'oneSentence', 'short', 'medium', 'long'.
 * Default is 'oneSentence'.
 *
 * @since 3.2.0
 * @hook classifai_azure_language_summary_length
 * @param {string} $summary_length The summary length.
 * @return {string} The summary length
 */
```

Sidsector9 commented on includes/Classifai/Providers/Azure/Language.php:
```diff
@@ -0,0 +1,391 @@
+				$response = json_decode( wp_remote_retrieve_body( $request ) );
+			}
+
+			$summary = $response->tasks->items[0]->results->documents[0]->summaries[0]->text;
```
Just want to confirm that the array indexes at `[0]` will always contain those properties, or if we should check via `isset()` and `count() > 0`?

psorensen committed:
update filter docblock

sksaju committed:
Make plugin VIP compliant by removing usage of wp_remote_get

sksaju committed:
address CR feedback on fix/791-make-plugin-vip-compliant-by-removing-usage-of-wp_remote_get

dkotter committed:
Merge pull request #901 from 10up/fix/791-make-plugin-vip-compliant-by-removing-usage-of-wp_remote_get

Fix | Make plugin VIP compliant by removing usage of wp_remote_get

psorensen committed:
Merge branch 'develop' into feature/azure-language-service

psorensen committed:
fix typo

psorensen committed:
phpcs

psorensen committed:
update text domain

psorensen committed:
fix undefined warning

psorensen committed:
Merge branch 'develop' into feature/azure-language-service

psorensen committed:
update azure endpoint

psorensen committed:
fix: Azure Language auth - wp_remote_get doesn't support body, use wp_remote_post

psorensen committed:
adds azure language provider settings

psorensen committed:
remove legacy settings

psorensen committed:
prevent infinite attempts when service never completes

psorensen committed:
filter max attempts

psorensen committed:
remove test

psorensen committed:
feat: add conditional rendering for Azure Language in excerpt generation settings

Hide prompt and length fields when Azure Language provider is selected since it uses predefined string options for these settings.

psorensen committed:
add additional error handling

psorensen committed:
Merge branch 'develop' into feature/azure-language-service

psorensen committed:
phpcs

psorensen committed:
lint js

dkotter committed:
Fix eslint error

psorensen committed:
Merge branch 'develop' into feature/azure-language-service

Commit:
remove duplicate function

Commit:
Merge branch 'develop' into feature/azure-language-service

Commit:
resolve duplicate use functions

Commit:
phpcs

Commit:
restore missing function

jeffpaul commented:
@dkotter @Sidsector9 back to you for review

jeffpaul commented:
Let's hold this for the 4.0.0 release where we may have things timed that if the WP AI Client gets into WordPress 7.0 that we could publish this as a composer package, add test coverage, and ensure this covers all respective features and not just excerpt generation.
